cpp/src/arrow/util/async_nursery_test.cc

```diff
@@ -0,0 +1,184 @@
+    AddDependentTask(std::move(close_future));
+  }
+
+  Future<> DoClose() override { return Future<>::MakeFinished(); }
+};
+
+class MarkWhenDestroyed : public GatingDoClose {
+ public:
+  MarkWhenDestroyed(Future<> close_future, bool* destroyed)
+      : GatingDoClose(std::move(close_future)), destroyed_(destroyed) {}
+  ~MarkWhenDestroyed() { *destroyed_ = true; }
+
+ private:
+  bool* destroyed_;
+};
+
+class EvictsChild : public AsyncCloseable {
+ public:
+  EvictsChild(Nursery* nursery, bool* child_destroyed, Future<> child_future,
+              Future<> final_future)
+      : final_close_future_(std::move(final_future)) {
+    owned_child_ = nursery->MakeSharedCloseable<MarkWhenDestroyed>(
+        std::move(child_future), child_destroyed);
+  }
+
+  void EvictChild() { owned_child_.reset(); }
+
+  Future<> DoClose() override { return final_close_future_; }
+
+ private:
+  Future<> final_close_future_;
+  std::shared_ptr<GatingDoClose> owned_child_;
+};
+
+template <typename T>
+void AssertDoesNotCloseEarly() {
+  Future<> gate = Future<>::Make();
+  std::atomic<bool> finished{false};
+  std::thread thread([&] {
+    ASSERT_OK(Nursery::RunInNursery(
+        [&](Nursery* nursery) { nursery->MakeSharedCloseable<T>(gate); }));
+    finished.store(true);
+  });
+
+  SleepABit();
+  ASSERT_FALSE(finished.load());
+  gate.MarkFinished();
+  BusyWait(10, [&] { return finished.load(); });
+  thread.join();
+}
+
+template <typename T>
+void AssertDoesNotCloseEarlyWithChild() {
+  Future<> gate = Future<>::Make();
+  std::atomic<bool> finished{false};
+  std::thread thread([&] {
+    ASSERT_OK(Nursery::RunInNursery(
+        [&](Nursery* nursery) { nursery->MakeSharedCloseable<T>(nursery, gate); }));
+    finished.store(true);
+  });
+
+  SleepABit();
+  ASSERT_FALSE(finished.load());
+  gate.MarkFinished();
+  BusyWait(10, [&] { return finished.load(); });
+  thread.join();
+}
+
+TEST(AsyncNursery, DoClose) { AssertDoesNotCloseEarly<GatingDoClose>(); }
+
+TEST(AsyncNursery, SharedChildDoClose) {
+  AssertDoesNotCloseEarlyWithChild<GatingDoCloseWithSharedChild>();
+}
+
+TEST(AsyncNursery, UniqueChildDoClose) {
+  AssertDoesNotCloseEarlyWithChild<GatingDoCloseWithUniqueChild>();
+}
+
+TEST(AsyncNursery, DependentTask) {
+  AssertDoesNotCloseEarly<GatingDoCloseAsDependentTask>();
+}
+
+TEST(AsyncNursery, EvictedChild) {
+  Future<> child_future = Future<>::Make();
+  Future<> final_future = Future<>::Make();
+  std::atomic<bool> finished{false};
+  std::thread thread([&] {
+    ASSERT_OK(Nursery::RunInNursery([&](Nursery* nursery) {
+      bool child_destroyed = false;
+      std::shared_ptr<EvictsChild> evicts_child =
+          nursery->MakeSharedCloseable<EvictsChild>(nursery, &child_destroyed,
+                                                    child_future, final_future);
+      evicts_child->EvictChild();
+      // Owner no longer has reference to child here but it's kept alive by nursery
+      // because it isn't done
```

Comment:
I don't understand this comment, because by reading the source code, I get the impression that the nursery doesn't keep anything alive (it does not have a container of tasks or futures).